oak_runtime/src/node/http/client.rs

```diff
@@ -0,0 +1,327 @@
+        // Create a [`Receiver`] to receive the HTTP request.
+        let receiver = Receiver::<Invocation>::new(ReadHandle { handle });
+        loop {
+            debug!("Waiting for HTTP invocation");
+            // Read an HTTP invocation from the [`Receiver`].
+            let invocation = receiver.receive(&runtime).map_err(|error| {
+                match error {
+                    OakError::OakStatus(OakStatus::ErrTerminated) => {
+                        debug!("HTTP client node is terminating.")
+                    }
+                    OakError::OakStatus(OakStatus::ErrChannelClosed) => {
+                        info!("HTTP invocation channel closed")
+                    }
+                    _ => error!("Couldn't receive the invocation: {:?}", error),
+                }
+                error
+            })?;
+
+            self.process_invocation(&runtime, &invocation).await;
+            debug!("Invocation processing finished");
+            invocation.close(&runtime);
+        }
+    }
+
+    /// Process an HTTP invocation for an external HTTP service.
+    async fn process_invocation(&mut self, runtime: &RuntimeProxy, invocation: &Invocation) {
+        // TODO(#1717): Update HTTP client metrics
+
+        // A closure to ease sending and logging the errors
+        let send_error = |err, code, msg| {
+            warn!("{}: {}", msg, err);
+            invocation.send_error(code, msg, runtime);
+        };
+
+        // Receive a request from the invocation channel.
+        let request = match invocation.receive_request(&runtime) {
+            Ok(request) => request,
+            Err(err) => {
+                send_error(
+                    format!("{:?}", err),
+                    http::StatusCode::INTERNAL_SERVER_ERROR,
+                    "Failed to read request",
+                );
+                return;
+            }
+        };
+        debug!("Incoming HTTP request: {:?}", request);
+
+        let mut hyper_request_builder = hyper::Request::builder()
+            .method(&request.method[..])
+            .uri(request.uri);
+
+        if let Some(headers) = request.headers {
+            for (header_name, header_value) in headers.into_iter() {
+                hyper_request_builder = hyper_request_builder.header(header_name, header_value);
+            }
+        }
+        let hyper_request = match hyper_request_builder.body(hyper::Body::from(request.body)) {
+            Ok(request) => request,
+            Err(err) => {
+                send_error(
+                    format!("{:?}", err),
+                    http::StatusCode::INTERNAL_SERVER_ERROR,
+                    "Failed to convert the request body",
+                );
+                return;
+            }
+        };
+
+        // Check that the `host` in the request URI matches the authority supported by the node, if
+        // any.
+        if let Some(authority) = self.authority.clone() {
```

Comment:
Conceptually, it feels like this check should happen as early as possible (as for most checks), especially since it's only based on the URI of the request. Everything else should happen only if this check was successful. For now it is not a big difference, but I am worried in the future the logic may grow, and some side-effects may accidentally sneak in before the validation is performed.